Mathlib/Analysis/Normed/Ring/SmoothingSeminorm.lean

```diff
@@ -0,0 +1,604 @@
+private theorem smoothingSeminormSeq_hasLimit_m (x : R) {ε : ℝ} (hε : 0 < ε) :
+    ∃ m : PNat, f (x ^ (m : ℕ)) ^ (1 / m : ℝ) <
+        (iInf fun n : PNat => f (x ^ (n : ℕ)) ^ (1 / (n : ℝ))) + ε / 2 :=
+  exists_lt_of_ciInf_lt (lt_add_of_le_of_pos (le_refl _) (half_pos hε))
+
+private theorem smoothingSeminormSeq_hasLimit_aux {L : ℝ} (hL : 0 ≤ L) {ε : ℝ} (hε : 0 < ε)
+    {m1 : ℕ} (hm1 : 0 < m1) {x : R} (hx : f x ≠ 0) :
+    Tendsto
+      (fun n : ℕ => (L + ε) ^ (-(((n % m1 : ℕ) : ℝ) / (n : ℝ))) * (f x ^ (n % m1)) ^ (1 / (n : ℝ)))
+      atTop (𝓝 1) := by
+  rw [← mul_one (1 : ℝ)]
+  have h_exp : Tendsto (fun n : ℕ => ((n % m1 : ℕ) : ℝ) / (n : ℝ)) atTop (𝓝 0) :=
+    tendsto_mod_div_atTop_nhds_zero_nat hm1
+  apply Tendsto.mul
+  · have h0 : Tendsto (fun t : ℕ => -(((t % m1 : ℕ) : ℝ) / (t : ℝ))) atTop (𝓝 0) := by
+      rw [← neg_zero]
+      exact Tendsto.neg h_exp
+    rw [← rpow_zero (L + ε)]
+    apply Tendsto.rpow tendsto_const_nhds h0
+    rw [ne_eq, add_eq_zero_iff_of_nonneg hL (le_of_lt hε)]
+    exact Or.inl (not_and_of_not_right _ (ne_of_gt hε))
+  · simp_rw [mul_one, ← rpow_natCast, ← rpow_mul (apply_nonneg f x), ← mul_div_assoc, mul_one, ←
+      rpow_zero (f x)]
+    exact Tendsto.rpow tendsto_const_nhds h_exp (Or.inl hx)
+
+/-- `smoothingSeminormSeq` is bounded below by zero. -/
+theorem smoothingSeminormSeq_bddBelow (x : R) :
+    BddBelow (Set.range fun n : ℕ+ => f (x ^ (n : ℕ)) ^ (1 / (n : ℝ))) := by
+  use 0
+  rintro y ⟨n, rfl⟩
+  exact rpow_nonneg (apply_nonneg f _) _
+
+/-- The iInf of the sequence `f(x^(n : ℕ)))^(1/(n : ℝ)`. -/
+def smoothingSeminorm' (x : R) : ℝ :=
+  iInf fun n : PNat => f (x ^ (n : ℕ)) ^ (1 / (n : ℝ))
+
+/-- If `f x = 0`, then `smoothingSeminorm' f x` is the limit of `smoothingSeminormSeq f x`. -/
+theorem tendsto_smoothingSeminorm'_of_eq_zero {x : R} (hx : f x = 0) :
+    Tendsto (smoothingSeminormSeq f x) atTop (𝓝 (smoothingSeminorm' f x)) := by
+  have h0 (n : ℕ) (hn : 1 ≤ n) : f (x ^ n) ^ (1 / (n : ℝ)) = 0 := by
+    have hfn : f (x ^ n) = 0 := by
+      apply le_antisymm _ (apply_nonneg f _)
+      rw [← zero_pow (pos_iff_ne_zero.mp hn), ← hx]
+      exact map_pow_le_pow _ x (one_le_iff_ne_zero.mp hn)
+    rw [hfn, zero_rpow (one_div_cast_ne_zero (one_le_iff_ne_zero.mp hn))]
+  have hL0 : (iInf fun n : PNat => f (x ^ (n : ℕ)) ^ (1 / (n : ℝ))) = 0 :=
+    le_antisymm
+      (ciInf_le_of_le (smoothingSeminormSeq_bddBelow f x) (1 : PNat) (le_of_eq (h0 1 (le_refl _))))
+      (le_ciInf fun n => rpow_nonneg (apply_nonneg f _) _)
+  simp only [hL0, smoothingSeminormSeq, smoothingSeminorm']
+  exact tendsto_atTop_of_eventually_const h0
+
+/-- If `f 1 ≤ 1` and `f x ≠  0`, then `smoothingSeminorm' f x` is the limit of
+`smoothingSeminormSeq f x`. -/
+theorem tendsto_smoothingSeminorm'_of_ne_zero (hf1 : f 1 ≤ 1) {x : R} (hx : f x ≠ 0) :
+    Tendsto (smoothingSeminormSeq f x) atTop (𝓝 (smoothingSeminorm' f x)) := by
```

Comment:
Can you add a bit of explanation for this long proof?

Reply:
I have added comments inside the proof.